internal/connector/connector.go

```diff
@@ -320,37 +318,47 @@ func Run(ctx context.Context, options Options) error {
 		options.Name = fmt.Sprintf("kubernetes.%s", options.Name)
 	}
 
-	serverName := "infra-connector"
+	caCertPEM, err := os.ReadFile(options.CACert)
+	if err != nil {
+		return err
+	}
 
-	manager := &autocert.Manager{
-		Prompt: autocert.AcceptTOS,
-		Cache:  autocert.DirCache(options.TLSCache),
+	caKeyPEM, err := os.ReadFile(options.CAKey)
+	if err != nil {
+		return err
 	}
 
-	var tlsConfig *tls.Config
+	ca, err := tls.X509KeyPair(caCertPEM, caKeyPEM)
+	if err != nil {
+		return err
+	}
 
-	if options.TLSCert != "" || options.TLSKey != "" {
-		certBytes, err := ioutil.ReadFile(options.TLSCert)
-		if err != nil {
-			return err
-		}
+	caCert, err := x509.ParseCertificate(ca.Certificate[0])
+	if err != nil {
+		return err
+	}
 
-		keypair, err := tls.LoadX509KeyPair(options.TLSCert, options.TLSKey)
-		if err != nil {
-			return err
-		}
+	caKey := ca.PrivateKey
 
-		tlsConfig = &tls.Config{
-			MinVersion:   tls.VersionTLS12,
-			Certificates: []tls.Certificate{keypair},
-		}
+	// validServernames stores which servernames Infra can generate certificates
+	// for to avoid generating certificates for any host
+	var validServernames sync.Map
+	validServernames.Store("", true)
+	validServernames.Store("localhost", true)
+	validServernames.Store("127.0.0.1", true)
+	validServernames.Store("infra-connector", true)
 
-		if err := manager.Cache.Put(context.TODO(), serverName, certBytes); err != nil {
-			return err
+	// Generate TLS certificates on the fly for clients
+	// GenerateCertificate caches certificates
+	tlsConfig := &tls.Config{MinVersion: tls.VersionTLS12}
+	tlsConfig.GetCertificate = func(hello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+		servername := hello.ServerName
```

Comment:
servername might be missing if client doesn't support it

Reply:
Great catch! Updated since the prior approach wouldn't work for IP addresses